### PR TITLE
Feature: CQRS pattern and event sourcing

### PR DESCRIPTION
**Pull request by KGibb8**

### CODE REVIEW

File: app/workers/asset_removed_from_portfolio.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Workers
+  class AssetRemovedFromPortfolio < ApplicationJob
+    queue_as :default
+
+    def perform(*args)
+      call(YAML.load(args.first))
```
Comment (nikolai-b) [inline app/workers/asset_removed_from_portfolio.rb]: I'm lost, I don't see any `YAML`....

File: app/workers/asset_added_to_portfolio.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    def find_portfolio
+      @portfolio = ::Portfolio.find_by(portfolio_attributes)
+    end
+
+    def create_portfolio
+      @portfolio = ::Portfolio.new(portfolio_attributes.merge(state: :draft)).tap(&:save)
+    end
+
+    def find_asset
+      @asset = ::Asset.find_by(asset_attributes.except(:quantity))
```
Comment (nikolai-b) [inline app/workers/asset_added_to_portfolio.rb]: I don't really follow this event sourcing idea.  It seems like you have an event stream but then we persist things to the database which then will change the state, i.e. how would you rewind an event or what if the worker changed then the meaning of the event would change?

File: config/initializers/active_record.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+Rails.application.configure do
+  config.generators do |g|
+    g.orm :active_record, primary_key_type: :uuid
```
Comment (nikolai-b) [inline config/initializers/active_record.rb]: nice

File: app/command/execute.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    end
+
+    private
+
+    def handler_for(command)
+      {
+        Command::AddAssetToPortfolio => CommandHandler::AddAssetToPortfolio.new,
+        Command::AdjustAssetValue => CommandHandler::AdjustAssetValue.new,
+        Command::FinalisePortfolio => CommandHandler::FinalisePortfolio.new
+      }.fetch(command.class)
```
[the remark this answers is not in the document]
Comment (KGibb8) [reply app/command/execute.rb]: You mean something along the lines of...

```
module Command
  class Base
    include ActiveModel::Model
    include ActiveModel::Validations
    include ActiveModel::Conversion

    def initialize(attributes = {})
      super
    end

    def validate!
      raise ValidationError, errors unless valid?
    end

    def persisted?
      false
    end
 
    def handler_class
      "CommandHandler::#{self.class.name.demodulize}".constantize
    end
  end
end
```

**Comment by KGibb8** [comment]
I've been using this [example repository](https://github.com/mpraglowski/cqrs-es-sample-with-res) to help get an idea of how to implement it. I do think that my approach at the moment isn't correct, it was more to do with getting familiar with the gem and how it works, and making that public so others could see. I have no idea how the sequence of events should play out at the moment. We need to discuss this.

**Comment by KGibb8** [comment]
Creating a new feature branch for this with a more thought out approach. Watch this space.

**Pull request closed without merging**